### PR TITLE
Allow opening an IPA bundle file directly

### LINKED ISSUE #20: Add support for opening IPA files
Please add support for opening IPA files directly, using the 7zip or a normal zip engine to extract it and run it.

### PR DESCRIPTION
**Pull request by DCNick3**
Closes #20

It works, but I am not yet happy about all the code. 
The error handling is kind of all over the place.
Also, maybe it's worth making a `FileRef` enum representing "a file that can potentially be opened" instead of two variants in `FsNode` enum

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -191,6 +266,49 @@ dependencies = [
...
+name = "crossbeam-utils"
+version = "0.8.14"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "4fb766fa798726286dbbb842f174001dab8abc7b627a1dd86e0b7222a95d929f"
+dependencies = [
+ "cfg-if",
+]
+
+[[package]]
+name = "crypto-common"
```
Comment (hikari-no-yume) [inline Cargo.lock]: Hmm, I assume this is being pulled in because of default features in the `zip` crate, presumably to support ZIP files with encryption, but I don't think we need it for IPA files? There's a few other new dependencies here that are presumably also not needed. Can you try cutting down the feature list for this dependency to just an empty one?

File: src/fs.rs [hunk 2]
```diff
@@ -91,20 +100,23 @@ impl FsNode {
...
+    }
 }
 
 /// Like [Path] but for the virtual filesystem.
 #[repr(transparent)]
 #[derive(Debug)]
 pub struct GuestPath(str);
 impl GuestPath {
-    pub fn new<S: AsRef<str>>(s: &S) -> &GuestPath {
+    pub fn new<S: AsRef<str> + ?Sized>(s: &S) -> &GuestPath {
```
Comment (hikari-no-yume) [inline src/fs.rs:120]: Oh, this is surprising to me because I thought this code worked. Was this constructor not working because I failed to include this bound? This looks fine though, I'm just wondering.

File: src/fs.rs [hunk 3]
```diff
@@ -276,6 +300,64 @@ fn handle_open_err<T>(open_result: std::io::Result<T>, host_path: &Path) -> T {
...
+}
+
+impl File {
+    fn from_host_file(file: std::fs::File) -> File {
+        File::HostFile(file)
+    }
+
+    fn from_ipa_file(file: IpaFile) -> File {
+        File::IpaBundleFile(file)
+    }
```
Comment (hikari-no-yume) [inline src/fs.rs]: A bit of a nitpick I guess, but since these aren't public, maybe we should just create the enum members directly for the time being? The only reason I have constructor functions for the `FsNode`s is that they helped a lot with the readability in `Fs::new`.

Actually, I notice that all the uses of `from_ipa_file` are paired with `file.open()` right now, so perhaps you could move that bit into this constructor. Then it would be justifying its own existence.

File: src/fs.rs [hunk 4]
```diff
@@ -276,6 +300,64 @@ fn handle_open_err<T>(open_result: std::io::Result<T>, host_path: &Path) -> T {
     }
 }
 
+/// Like [std::fs::File] but for the guest filesystem.
+#[derive(Debug)]
+pub enum File {
```
Comment (hikari-no-yume) [inline src/fs.rs]: Could you rename this to `GuestFile`? I try to avoid using the same names as well-known `std` types, especially if they're conceptually very similar, so that they'll be easy to distinguish if we import both in the same file. Hence `GuestPath`, `GuestUSize`, etc.

If you do, you can then add `use std::fs::File;` at the top of this file, if you feel like it.

File: src/fs.rs [hunk 5]
```diff
@@ -418,40 +494,40 @@ impl Fs {
 
     /// Like [std::path::Path::is_file] but for the guest filesystem.
     pub fn is_file(&self, path: &GuestPath) -> bool {
-        matches!(self.lookup_node(path), Some(FsNode::File { .. }))
+        matches!(self.lookup_node(path), Some(FsNode::HostFile { .. }))
```
Comment (hikari-no-yume) [inline src/fs.rs]: Uhh, this looks like a bug, we want files from the IPA bundle to be considered “files” too, right?

File: .gitignore [hunk 6]
```diff
@@ -1,3 +1,4 @@
 /target
 /vendor/boost
 /touchHLE_sandbox
+.idea
```
Comment (hikari-no-yume) [inline .gitignore]: Was this added automatically by your IDE? I think there's some way to tell git to ignore files locally, without having to modify the repo.

File: src/fs.rs [hunk 7]
```diff
@@ -276,6 +300,64 @@ fn handle_open_err<T>(open_result: std::io::Result<T>, host_path: &Path) -> T {
...
+            File::IpaBundleFile(file) => file.read(buf),
+        }
+    }
+}
+
+impl Write for File {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        match self {
+            File::HostFile(file) => file.write(buf),
+            File::IpaBundleFile(file) => file.write(buf),
```
Comment (hikari-no-yume) [inline src/fs.rs]: I think we'll always want to treat the IPA file as read-only. I see you have a `Write` implementation in `src/fs/bundle.rs`, but maybe you could remove it and just panic here instead.

File: src/fs.rs [hunk 8]
```diff
@@ -418,40 +494,40 @@ impl Fs {
...
+        let mut result = Vec::new();
+        file.read_to_end(&mut result).map_err(|_| ())?;
+        Ok(result)
     }
 
     /// Like [std::fs::File::open] but for the guest filesystem.
     #[allow(dead_code)]
-    pub fn open<P: AsRef<GuestPath>>(&self, path: P) -> Result<std::fs::File, ()> {
+    pub fn open<P: AsRef<GuestPath>>(&self, path: P) -> Result<File, ()> {
+        // it would be nice to delegate to self.open_with_options, but currently it wants a mutable reference to self
```
Comment (hikari-no-yume) [inline src/fs.rs:519]: Ah, so you ran into this too, that's indeed exactly why I have the two implementations. I'm a bit scared they'll diverge in behaviour in some corner case eventually, but not requiring `&mut` is nice. Maybe I should write some tests for this code, it's some of the trickiest stuff to get right and also the easiest to test.

(Just dismiss this, I'm not asking you to change the comment, it's good!)

File: src/fs/bundle.rs [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+impl FsNodeBuilder {
+    fn new() -> Self {
+        Self {
+            root: FsNode::dir(),
+        }
+    }
+
+    fn find_or_make_directory(&mut self, path: &GuestPath) -> &mut FsNode {
```
Comment (hikari-no-yume) [inline src/fs/bundle.rs:29]: Just curious, is this because the entry for `/SomeDir/SomeFoo` can appear before the entry for `/SomeDir` or something like that?

File: src/fs/bundle.rs [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+            let next = children
+                .entry(part.to_string())
+                .or_insert_with(|| FsNode::dir());
+            current = next;
+        }
+        current
+    }
+
+    fn add_file(&mut self, path: &GuestPath, node: FsNode) {
+        // I __hope__ no adversory will try to create files with .. in their names...
```
Comment (hikari-no-yume) [inline src/fs/bundle.rs]: Maybe `assert!` just in case. (FWIW I think the right spelling is `adversary`.)

File: src/fs/bundle.rs [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    Zip {
+        zip: ZipArchive<std::fs::File>,
+        /// Path to the app bundle inside the zip file.
+        /// It should be "Payload/<app name>.app" (no trailing slash!).
+        bundle_path: String,
+    },
+}
+
+impl BundleData {
+    // TODO actually, the error type is not at all a ZipError.. I would like to use a custom error type here (anyhow, snafu, thiserror, ...), but there doesnt seem to be one used in the project?
```
Comment (hikari-no-yume) [inline src/fs/bundle.rs]: Maybe you can just use a `String` or something? I don't think there's going to be any interesting context, we'll either find the bundle or we won't.

File: src/fs/bundle.rs [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    fn find_bundle_path_in_archive(
+        zip: &mut ZipArchive<std::fs::File>,
+    ) -> Result<String, ZipError> {
+        static BUNDLE_NAME_REGEX: Lazy<Regex> =
+            Lazy::new(|| Regex::new(r"Payload/([^/]+)\.app/").unwrap());
```
Comment (hikari-no-yume) [inline src/fs/bundle.rs]: It's not a big deal, but I think we could match this pattern without needing a regex library.

File: src/fs/bundle.rs [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+                Ok(buf)
+            }
+        }
+    }
+}
+
+/// Represents a file inside an IPA bundle that can be opened.
+#[derive(Debug)]
+pub struct IpaFileRef {
+    archive: Arc<Mutex<ZipArchive<std::fs::File>>>,
```
Comment (hikari-no-yume) [inline src/fs/bundle.rs]: Currently this project is _very_ single-threaded, even though it simulates a potentially multi-threaded app, so it'd be fine to use `Rc` instead. I assume we can easily change it back in future? I'm not sure about when it's necessary to use a `Mutex` in Rust, but I assume it has a single-threaded replacement too.

File: src/fs/bundle.rs [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        }
+    }
+}
+
+/// Represents an opened file in an IPA bundle.
+pub struct IpaFile {
+    // we need to use a cursor because zip::read::ZipFile doesn't implement Seek
+    // and, generally, seeking in compressed files is hard to achieve
+    // the simplest way to do it is to read the whole file into memory
+    // the target apps should be small enough to fit in memory, right?
```
Comment (hikari-no-yume) [inline src/fs/bundle.rs:210]: Hmm, this might end up a problem with large files (app bundles can contain video files and stuff), but I assume they won't be kept open for too long so it sounds fine for now.

**Review by hikari-no-yume — COMMENTED** [review]
Hey, thanks a lot! This looks pretty great in general, I appreciate the way you've abstracted things, I'd have done something similar. I didn't look very closely at the new `bundle.rs` file, but I assume it's fine.

I have various comments. If you don't feel like doing all of them, I'm happy to do some of them myself.

File: src/fs.rs (hunk 2 above)
Comment (DCNick3) [reply src/fs.rs:120]: Yeah, you could not call this constructor with `&str` argument

File: src/fs.rs (hunk 5 above)
Comment (DCNick3) [reply src/fs.rs]: Ah, I knew I missed something! This is why I want a "FileRef" enum variant

File: .gitignore (hunk 6 above)
Comment (DCNick3) [reply .gitignore]: It was added by me, because CLion puts some stuff in `.idea` (and if somebody will try to open this in CLion they may accidentally commit some garbage). I can remove it and put it to my local .gitignore if you want though

File: src/fs.rs (hunk 8 above)
Comment (DCNick3) [reply src/fs.rs:519]: It might be a good idea to have interior mutability (like a `Mutex` or a `Cell`) in fs, so it would always take a `&self`?

File: .gitignore (hunk 6 above)
Comment (hikari-no-yume) [reply .gitignore]: I'd prefer that since this might end up being a laundry list of IDEs otherwise, which is a distraction from the main purpose IMO.

File: src/fs.rs (hunk 2 above)
Comment (hikari-no-yume) [reply src/fs.rs:120]: Huh, not sure how I missed that. I wasn't aware of this detail of Rust, thanks for the fix!

File: .gitignore (hunk 6 above)
Comment (DCNick3) [reply .gitignore]: >  laundry list of IDEs

It often is, just look at https://github.com/dotnet/runtime/blob/main/.gitignore

File: src/fs.rs (hunk 5 above)
Comment (hikari-no-yume) [reply src/fs.rs]: Maybe not in this in-line comment section, but if you could elaborate a bit more on why you want it, that'd be great.

File: src/fs/bundle.rs (hunk 9 above)
Comment (DCNick3) [reply src/fs/bundle.rs:29]: Yeah, I think this is it. The zip spec says "files can come in any order", extraction logic in `zip` crate always tries to create parent directories, so I guess this can be possible (even though it's weird)

File: src/fs/bundle.rs (hunk 13 above)
Comment (DCNick3) [reply src/fs/bundle.rs]: Hmmm, how are gonna implement threading btw? I tried to design a threading system in my HLE and found it... Non-trivial, so to say

File: .gitignore (hunk 6 above)
Comment (hikari-no-yume) [reply .gitignore]: I guess it's a good idea if you have a large project and a lot of novice contributors, but I feel like it's a workaround for people not being familiar with better tools, to be honest. I'd rather keep the `.gitignore` to stuff specific to this project.

File: src/fs/bundle.rs (hunk 14 above)
Comment (DCNick3) [reply src/fs/bundle.rs:210]: Yeah, I figured this too. Just something to be aware of

File: src/fs/bundle.rs (hunk 12 above)
Comment (DCNick3) [reply src/fs/bundle.rs]: I don't want to reinvent regex =(

File: src/fs/bundle.rs (hunk 11 above)
Comment (DCNick3) [reply src/fs/bundle.rs]: Yeah, based on discussion in #33, I think it's fine to use `String` here

File: src/fs.rs (hunk 8 above)
Comment (hikari-no-yume) [reply src/fs.rs:519]: I think I would like it if you need `&mut` to “mutate” the FS somehow, but interior mutability would be a nice trick to solve the implementation-sharing problem. We can enforce that invariant within a wrapper function in that case. Changing this should be a separate commit or PR though.

File: Cargo.lock (hunk 1 above)
Comment (DCNick3) [reply Cargo.lock]: Just an empty one won't do (all the compression formats are features :P), but yeah, crypto is probably not something we touchHLE needs

File: src/fs/bundle.rs [hunk 15]
```diff
@@ -0,0 +1,235 @@
+use crate::fs::{FsNode, GuestPath};
```
Comment (hikari-no-yume) [inline src/fs/bundle.rs:7]: Oh right, can you add the MPL-2.0 license header here (please copy it from another file), and a doc comment (a one-liner is fine) summarising what this file is for? The license header is necessary because MPL-2.0 is a per-file rather than whole-project license, so it needs to be explicit which files it applies to, and the doc comments are just a thing I like to have to make the project nice to navigate in `cargo doc`.

File: src/fs.rs (hunk 8 above)
Comment (DCNick3) [reply src/fs.rs:519]: Sure

File: Cargo.lock (hunk 1 above)
Comment (hikari-no-yume) [reply Cargo.lock]: Oh haha, I assume it's just DEFLATE we need then? (I haven't actually checked…)

Comment (DCNick3) [reply Cargo.lock]: Deflate is definitely the most popular zip compression method. I can imagine some files using a different compression, but.. Compile times require sacrifices or smth

File: src/fs/bundle.rs (hunk 12 above)
Comment (hikari-no-yume) [reply src/fs/bundle.rs]: That's… fair. I just think Rust usually offers enough nice ways to do things (e.g. the `str` methods and `Iter`) that regex is a bit overkill, and I'm a bit wary of it being deceptively simple sometimes (e.g. unclear greedy vs non-greedy matching behaviour).

I think something like this would work?

```rust
if let Some(name) = path
    .strip_prefix("Payload/")
    .and_then(|path| path.split_once('/'))
    .and_then(|(name, _)| name.strip_suffix(".app")) {
```

File: src/fs/bundle.rs (hunk 13 above)
Comment (hikari-no-yume) [reply src/fs/bundle.rs]: There's already a very primitive threading system since Super Monkey Ball needs it. I can explain a bit about how it works in some other channel (I don't want to get too off-topic here), but you can see the guts of it in `main.rs`. Incidentally, I really should move it out of that file, it's getting too complicated.

**Comment by hikari-no-yume** [comment]
Oh right, the README and maybe the help/usage text should be updated after this. I'm happy to do that myself, I like writing documentation.

**Comment by DCNick3** [comment]
I think all the comments are addressed now

**Comment by DCNick3** [comment]
Regarding the `FileRef`: When I first added the `IpaBundleFile` enum variant, I completely forgot to update the places where matches are made over it (and the `is_file` even made it into git).

What I think could have helped is to add another level of indirection (as always): have a "this is a file" enum variant in the `FsNode`  and an inner `FileRef` enum specifying which kind. 

This way the `is_file` problem would not have happened (one enum variant for all files, no need to update)

In the places you actually need to look inside the file (like open), you would first match over the `FsNode` to check whether it's a file and then you would match __exhaustively__ over `FileRef` (as you no longer need "it's a directory" error) and, therefore, changing the `FileRef` enum will give you some compiler errors to tell you what needs updating.

But I don't think it's a good idea to it in this PR

File: src/fs/bundle.rs (hunk 11 above)
Comment (hikari-no-yume) [reply src/fs/bundle.rs]: Can you remove the comment?

Comment (hikari-no-yume) [reply src/fs/bundle.rs]: Wait I'm not sure if GitHub is showing me the latest changes, uhh…

**Comment by hikari-no-yume** [comment]
Yeah, that justification makes sense. I'll think about doing something like that if the enum gets any more complicated than it already is.

File: src/fs/bundle.rs (hunk 11 above)
Comment (DCNick3) [reply src/fs/bundle.rs]: It's "outdated". You need to find this place in the changes tab again ig. I think there is also a way to view changes from the last review.

image

File: src/fs.rs [hunk 16]
```diff
@@ -20,15 +20,25 @@
 //! Directories only need a corresponding directory in the host filesystem if
```
Comment (hikari-no-yume) [inline src/fs.rs:20]: Oh, the paragraph above this is outdated now that we have ZIP files. Uhh, I think I'll try to rewrite it myself later if that's okay.

Comment (DCNick3) [reply src/fs.rs:20]: Sure

File: src/main.rs [hunk 17]
```diff
@@ -282,10 +282,12 @@ impl Environment {
...
         let startup_time = std::time::Instant::now();
 
-        let (bundle, fs) = match bundle::Bundle::new_bundle_and_fs_from_host_path(bundle_path) {
+        let bundle_data = fs::BundleData::open_any(&bundle_path)
+            .map_err(|e| format!("Could not open app bundle: {e}"))?;
+        let (bundle, fs) = match bundle::Bundle::new_bundle_and_fs_from_host_path(bundle_data) {
             Ok(bundle) => bundle,
             Err(err) => {
-                return Err(format!("Application bundle error: {}. Check that the path is to a .app directory. If this is a .ipa file, you need to extract it as a ZIP file to get the .app directory.", err));
+                return Err(format!("Application bundle error: {err}. Check that the path is to a .app directory or an .ipa file."));
```
Comment (hikari-no-yume) [inline src/main.rs]: Super nitpick, I know, but: this usage of “a” versus “an” isn't internally consistent with the message :p

**Comment by hikari-no-yume** [comment]
Compiling this on my machine right now, I'll tell you once I've tested it a bit. I'll also have a look at the handful of IPA files I have to see what compression methods they have.

**Comment by hikari-no-yume** [comment]
`cargo clippy` has three suggestions for me now, maybe you'll want to address them. Otherwise I'll do it myself later.

File: src/dyld.rs [hunk 18]
```diff
@@ -123,7 +123,7 @@ pub fn search_lists<T>(
         .iter()
         .flat_map(|&n| n)
         .find(|&(sym, _)| *sym == symbol)
-        .map(|&(_, ref f)| f)
+        .map(|(_, f)| f)
```
Comment (hikari-no-yume) [inline src/dyld.rs]: Sorry, super nitpick, can you revert the change in this file? Just because it's nothing to do with your changes.

**Comment by hikari-no-yume** [comment]
I guess I have a newer Rust toolchain than you because there's one more lint for me:

```
warning: use of `ok_or` followed by a function call
  --> src/bundle.rs:36:14
   |
36 |             .ok_or("plist root value is not a dictionary".to_string())?;
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `ok_or_else(|| "plist root value is not a dictionary".to_string())`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
   = note: `#[warn(clippy::or_fun_call)]` on by default

warning: `touchHLE` (bin "touchHLE") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 10.52s
```

File: src/dyld.rs (hunk 18 above)
Comment (DCNick3) [reply src/dyld.rs]: Whoops :sweat_smile:, just run `cargo clippy --fix`

**Comment by DCNick3** [comment]
I've pushed the fixes for the lints

Just a note: in rust 1.66 a new [uninlined_format_args](https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args) lint was added, which is triggered in... almost the entire project :P (that's why I was kinda ignoring clippy here)

It's machine fixable, so probably want to update the toolchain & run `cargo clippy --fix`

**Comment by DCNick3** [comment]
> I guess I have a newer Rust toolchain than you because there's one more lint for me:

I am on latest stable, 1.67. Still, will fix it

**Comment by hikari-no-yume** [comment]
Ohh wow, I've stubbornly avoided using inlined format args, but maybe it's time for me to switch. Also, ah, for some reason I've never used `cargo clippy --fix`, I guess I've been wasting time :p

**Comment by DCNick3** [comment]
> I've stubbornly avoided using inlined format args

Same, lol. Now that clippy is complaining I have no choice ig

**Comment by hikari-no-yume** [comment]
macOS has an `unzip` utility, I assume the same or a similar tool is available on common Linux distros. I was able to write this one-liner to check a bunch of IPA files' compression methods:

```sh
for i in *.ipa; do unzip -v "$i" | cut -f 3,1 -w | grep -v 'Defl' | grep -v 'Stored' ; done
```

I ran this on all the IPAs I have on my drive: all the ones I redownloaded from the iTunes Store, and a number of others.

As expected, everything is either DEFLATE or uncompressed (aka “Stored”). I think the latter is maybe just for empty files?

So not supporting other compression methods seems pretty safe. 👍

**Comment by hikari-no-yume** [comment]
Oops, I got distracted. Works great in my testing. I was a bit worried it'd be slower than loading a pre-extracted file, but I can't notice the difference in release builds.

I think I'm ready to merge this. My preference is a squash-and-rebase merge type. Do you want to squash all the commits into one yourself (if so, please use `--reset-author` so the author date is more accurate) or shall I do it?

Thanks a lot for this contribution!

**Comment by DCNick3** [comment]
>I was a bit worried it'd be slower than loading a pre-extracted file, but I can't notice the difference in release builds.

The whole game is like 40 megabytes, really small for modern hardware

>Do you want to squash all the commits into one yourself

Sure

**Comment by hikari-no-yume** [comment]
Yeah I guess, but I know from one of my emulator developer friends that even small titles can chug if you have to decompress too often! Luckily Monkey Ball has loading screens.

**Comment by DCNick3** [comment]
(Did I do it right? never squashed this way before...)

**Comment by hikari-no-yume** [comment]
Yeah, looks good to me! My preference is to remove the info in the commit message about which commits were squashed, since that's usually useless extra info in future when reading the commit history, so I'll remove that bit when I cherry-pick the commit, if that's okay.

**Comment by hikari-no-yume** [comment]
Merged as 32d12cc697f8fc62770962b8d5a1fee44dfba677, thanks!

I added a changelog and credited you via your GitHub username in c407ef831b9f67ef0ced294dc30208108b16cf2c. If you'd prefer a different style of credit, please let me know.
